File: frontend/src/services/customerService.ts
import { apiService } from './api';
import { API_ENDPOINTS } from '../constants';
import { Customer, CustomerRequest, PagedResponse, ApiResponse } from '../types';

export interface CustomerSearchParams {
  query?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'ASC' | 'DESC';
}

export interface CustomerListParams {
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'ASC' | 'DESC';
}

/**
 * Customer Service API functions
 * All functions require ADMIN role as enforced by backend
 */
export const customerService = {
  /**
   * Get all customers with pagination
   */
  getCustomers: async (params: CustomerListParams = {}): Promise<PagedResponse<Customer>> => {
    const {
      page = 0,
      size = 20,
      sortBy = 'name',
      sortDirection = 'ASC'
    } = params;

    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sortBy,
      sortDirection
    });

    const response = await apiService.get<PagedResponse<Customer>>(
      `${API_ENDPOINTS.CUSTOMERS.BASE}?${queryParams}`
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch customers');
    }

    return response.data;
  },

  /**
   * Get customer by ID
   */
  getCustomerById: async (id: number): Promise<Customer> => {
    const response = await apiService.get<Customer>(
      API_ENDPOINTS.CUSTOMERS.BY_ID(id)
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch customer');
    }

    return response.data;
  },

  /**
   * Get customer by account number
   */
  getCustomerByAccountNumber: async (accountNumber: string): Promise<Customer> => {
    const response = await apiService.get<Customer>(
      API_ENDPOINTS.CUSTOMERS.BY_ACCOUNT(accountNumber)
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch customer');
    }

    return response.data;
  },

  /**
   * Create new customer
   */
  createCustomer: async (customerData: CustomerRequest): Promise<Customer> => {
    const response = await apiService.post<Customer>(
      API_ENDPOINTS.CUSTOMERS.BASE,
      customerData
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to create customer');
    }

    return response.data;
  },

  /**
   * Update existing customer
   */
  updateCustomer: async (id: number, customerData: CustomerRequest): Promise<Customer> => {
    const response = await apiService.put<Customer>(
      API_ENDPOINTS.CUSTOMERS.BY_ID(id),
      customerData
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to update customer');
    }

    return response.data;
  },

  /**
   * Delete customer
   */
  deleteCustomer: async (id: number): Promise<void> => {
    const response = await apiService.delete<string>(
      API_ENDPOINTS.CUSTOMERS.BY_ID(id)
    );

    if (!response.success) {
      throw new Error(response.message || 'Failed to delete customer');
    }
  },

  /**
   * Search customers
   */
  searchCustomers: async (params: CustomerSearchParams): Promise<PagedResponse<Customer>> => {
    const {
      query = '',
      page = 0,
      size = 20,
      sortBy = 'name',
      sortDirection = 'ASC'
    } = params;

    const queryParams = new URLSearchParams({
      query,
      page: page.toString(),
      size: size.toString(),
      sortBy,
      sortDirection
    });

    const response = await apiService.get<PagedResponse<Customer>>(
      `${API_ENDPOINTS.CUSTOMERS.SEARCH}?${queryParams}`
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to search customers');
    }

    return response.data;
  },

  /**
   * Get customers for auto-complete
   */
  getCustomersForAutoComplete: async (prefix: string): Promise<Customer[]> => {
    const queryParams = new URLSearchParams({ prefix });
    
    const response = await apiService.get<Customer[]>(
      `${API_ENDPOINTS.CUSTOMERS.AUTOCOMPLETE}?${queryParams}`
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch customer suggestions');
    }

    return response.data;
  },

  /**
   * Get customers without bills
   */
  getCustomersWithoutBills: async (): Promise<Customer[]> => {
    const response = await apiService.get<Customer[]>(
      API_ENDPOINTS.CUSTOMERS.WITHOUT_BILLS
    );

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch customers without bills');
    }

    return response.data;
  },

  /**
   * Get total customer count
   */
  getTotalCustomerCount: async (): Promise<number> => {
    const response = await apiService.get<number>(
      API_ENDPOINTS.CUSTOMERS.COUNT
    );

    if (!response.success || response.data === undefined) {
      throw new Error(response.message || 'Failed to fetch customer count');
    }

    return response.data;
  },

  /**
   * Validate customer data before submission
   */
  validateCustomerData: (data: CustomerRequest): string[] => {
    const errors: string[] = [];

    if (!data.name?.trim()) {
      errors.push('Customer name is required');
    } else if (data.name.length < 2 || data.name.length > 100) {
      errors.push('Name must be between 2 and 100 characters');
    }

    if (!data.address?.trim()) {
      errors.push('Address is required');
    } else if (data.address.length < 10 || data.address.length > 500) {
      errors.push('Address must be between 10 and 500 characters');
    }

    if (!data.phone?.trim()) {
      errors.push('Phone number is required');
    } else if (!/^(\+94|0)[0-9]{9}$/.test(data.phone)) {
      errors.push('Invalid Sri Lankan phone number format');
    }

    if (data.email && data.email.trim()) {
      if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(data.email)) {
        errors.push('Invalid email format');
      } else if (data.email.length > 100) {
        errors.push('Email must not exceed 100 characters');
      }
    }

    return errors;
  }
};

export default customerService;

File: backend/src/main/java/com/pahana/edu/bookshop/service/CustomerService.java
package com.pahana.edu.bookshop.service;

import com.pahana.edu.bookshop.dto.request.CustomerRequest;
import com.pahana.edu.bookshop.dto.response.CustomerResponse;
import com.pahana.edu.bookshop.entity.Customer;
import com.pahana.edu.bookshop.exception.BusinessException;
import com.pahana.edu.bookshop.exception.ResourceNotFoundException;
import com.pahana.edu.bookshop.repository.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

/**
 * Service class for Customer business logic
 * 
 * Handles customer management operations including CRUD operations,
 * validation, and business rule enforcement.
 */
@Service
@Transactional
public class CustomerService {

    @Autowired
    private CustomerRepository customerRepository;

    /**
     * Get all customers with pagination
     */
    @Transactional(readOnly = true)
    public Page<CustomerResponse> getAllCustomers(Pageable pageable) {
        Page<Customer> customers = customerRepository.findAll(pageable);
        return customers.map(this::convertToResponse);
    }

    /**
     * Get customer by ID
     */
    @Transactional(readOnly = true)
    public CustomerResponse getCustomerById(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", id));
        return convertToResponse(customer);
    }

    /**
     * Get customer by account number
     */
    @Transactional(readOnly = true)
    public CustomerResponse getCustomerByAccountNumber(String accountNumber) {
        Customer customer = customerRepository.findByAccountNumber(accountNumber)
                .orElseThrow(() -> new ResourceNotFoundException("Customer", "accountNumber", accountNumber));
        return convertToResponse(customer);
    }

    /**
     * Create new customer
     */
    public CustomerResponse createCustomer(CustomerRequest request) {
        // Validate unique constraints
        validateUniqueConstraints(request.getPhone(), request.getEmail(), null);
        
        // Create customer entity
        Customer customer = Customer.builder()
                .accountNumber(generateAccountNumber())
                .name(request.getName())
                .address(request.getAddress())
                .phone(request.getPhone())
                .email(request.getEmail())
                .build();

        Customer savedCustomer = customerRepository.save(customer);
        return convertToResponse(savedCustomer);
    }

    /**
     * Update existing customer
     */
    public CustomerResponse updateCustomer(Long id, CustomerRequest request) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", id));

        // Validate unique constraints (exclude current customer)
        validateUniqueConstraints(request.getPhone(), request.getEmail(), id);

        // Update fields
        customer.setName(request.getName());
        customer.setAddress(request.getAddress());
        customer.setPhone(request.getPhone());
        customer.setEmail(request.getEmail());

        Customer savedCustomer = customerRepository.save(customer);
        return convertToResponse(savedCustomer);
    }

    /**
     * Delete customer
     */
    public void deleteCustomer(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", id));

        // Check if customer has bills
        if (!customer.getBills().isEmpty()) {
            throw new BusinessException("Cannot delete customer with existing bills. " +
                    "Customer has " + customer.getBills().size() + " bills.");
        }

        customerRepository.delete(customer);
    }

    /**
     * Search customers
     */
    @Transactional(readOnly = true)
    public Page<CustomerResponse> searchCustomers(String searchTerm, Pageable pageable) {
        Page<Customer> customers = customerRepository.searchCustomers(searchTerm, pageable);
        return customers.map(this::convertToResponse);
    }

    /**
     * Get customers for auto-complete
     */
    @Transactional(readOnly = true)
    public List<CustomerResponse> getCustomersForAutoComplete(String prefix) {
        List<Customer> customers = customerRepository.findByNameStartingWithIgnoreCase(prefix);
        return customers.stream()
                .map(this::convertToResponse)
                .toList();
    }

    /**
     * Get customers without bills
     */
    @Transactional(readOnly = true)
    public List<CustomerResponse> getCustomersWithoutBills() {
        List<Customer> customers = customerRepository.findCustomersWithoutBills();
        return customers.stream()
                .map(this::convertToResponse)
                .toList();
    }

    /**
     * Get total customer count
     */
    @Transactional(readOnly = true)
    public long getTotalCustomerCount() {
        return customerRepository.getTotalCustomerCount();
    }

    /**
     * Generate unique account number
     */
    private String generateAccountNumber() {
        String accountNumber;
        do {
            // Generate account number format: CUST-YYYYMMDD-XXXX
            String datePart = java.time.LocalDate.now().toString().replace("-", "");
            String randomPart = UUID.randomUUID().toString().substring(0, 4).toUpperCase();
            accountNumber = "CUST-" + datePart + "-" + randomPart;
        } while (customerRepository.existsByAccountNumber(accountNumber));
        
        return accountNumber;
    }

    /**
     * Validate unique constraints
     */
    private void validateUniqueConstraints(String phone, String email, Long excludeId) {
        // Check phone number uniqueness
        if (phone != null && customerRepository.existsByPhone(phone)) {
            if (excludeId == null || !customerRepository.findByPhone(phone)
                    .map(Customer::getId).equals(excludeId)) {
                throw new BusinessException("Phone number already exists: " + phone);
            }
        }

        // Check email uniqueness
        if (email != null && customerRepository.existsByEmail(email)) {
            if (excludeId == null || !customerRepository.findByEmail(email)
                    .map(Customer::getId).equals(excludeId)) {
                throw new BusinessException("Email already exists: " + email);
            }
        }
    }

    /**
     * Convert Customer entity to CustomerResponse DTO
     */
    private CustomerResponse convertToResponse(Customer customer) {
        return CustomerResponse.builder()
                .id(customer.getId())
                .accountNumber(customer.getAccountNumber())
                .name(customer.getName())
                .address(customer.getAddress())
                .phone(customer.getPhone())
                .email(customer.getEmail())
                .createdAt(customer.getCreatedAt())
                .updatedAt(customer.getUpdatedAt())
                .totalBills(customer.getBills().size())
                .displayName(customer.getDisplayName())
                .build();
    }
}